src/instructlab/cli/model/upload.py

```diff
@@ -0,0 +1,239 @@
+    """Class to handle uploading safetensors and GGUF models to Hugging Face"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        destination: str,
+        release: str,
+        hf_token: str,
+    ) -> None:
+        super().__init__(
+            ctx=ctx,
+            model=model,
+            destination=destination,
+            release=release,
+        )
+        self.hf_token = (
+            hf_token.strip()
+        )  # Remove trailing whitespaces from the token if they exist
+        self.local_model_path = ""
+
+    def upload(self):
+        """
+        Upload specified model to Hugging Face
+        """
+
+        # HF token check
+        if len(self.hf_token) == 0:
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to upload HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
+
+        click.echo(
+            f"Uploading model to Hugging Face:\n{DEFAULT_INDENT}Model: {self.model}\n{DEFAULT_INDENT}Destination: {self.destination}@{self.release}\n"
+        )
+
+        # set logger
+        if self.ctx.obj is not None:
+            hf_logging.set_verbosity(self.ctx.obj.config.general.log_level.upper())
+
+        # determine if model is path or name - look in checkpoints dir if the latter
+        if os.path.exists(self.model):
+            self.local_model_path = Path(self.model)
+            self.model = self.model.split("/")[-1]
+        else:
+            self.local_model_path = os.path.join(DEFAULTS.CHECKPOINTS_DIR, self.model)
+            # throw an error if the built path is invalid
+            if not os.path.exists(self.local_model_path):
+                click.secho(
+                    f"Couldn't find model at {self.local_model_path} - are you sure it exists?",
+                    fg="red",
+                )
+                raise click.exceptions.Exit(1)
+
+        # upload either safetensors or gguf, if valid
+        if is_model_safetensors(self.local_model_path):
+            self.upload_safetensors()
+        elif is_model_gguf(self.local_model_path):
+            self.upload_gguf()
+        else:
+            click.secho(
+                f"Local model path {self.local_model_path} is a valid path, but is neither safetensors nor a GGUF - cannot upload to Hugging Face",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+    def upload_gguf(self) -> None:
+        try:
+            hf_api = HfApi()
+            resp = hf_api.upload_file(
+                path_or_fileobj=self.local_model_path,
+                path_in_repo=self.model,
+                repo_id=self.destination,
+                revision=self.release,
+                token=self.hf_token,
+            )
+            logger.debug(f"Hugging Face response:\n{resp}")
```

Comment:
sorry, not familiar with HfApi, is it able to check the exist model? Or it can repeat to upload?

Reply:
Here's the docs for it, we are using both `upload_file` and `upload_folder` https://huggingface.co/docs/huggingface_hub/v0.26.5/en/package_reference/hf_api#huggingface_hub.HfApi